src/shogun/multiclass/tree/CARTree.cpp

```diff
@@ -109,7 +109,30 @@ std::shared_ptr<MulticlassLabels> CARTree::apply_multiclass(std::shared_ptr<Feat
 	auto current=get_root()->as<bnode_t>();
 
 	require(current, "Tree machine not yet trained.");
-	return apply_from_current_node(data->as<DenseFeatures<float64_t>>(), current)->as<MulticlassLabels>();
+	if (auto tmp =
```

Comment:
maybe give this a more significant name, rather then tmp?
also use `->as<DenseSubSamplesFeatures<float64_t>>`

Reply:
use ->as<DenseSubSamplesFeatures<float64_t>> will give me a error. Because the implemention of as is
`template<class T>
	inline std::shared_ptr<T> as()
	{
		auto c = std::dynamic_pointer_cast<T>(shared_from_this());
		if (c)
			return c;

		error(
			"Object of type {} cannot be converted to type {}.",
			this->get_name(),
			demangled_type<T>().c_str());
		return nullptr;
	}`

Reply:
do you get a compile or runtime error?

Reply:
yes, compiler give a error that `"Object of type DenseFeatures cannot be converted to type DenseSubSamplesFeatures"`.

Reply:
You are calling the wrong function, I think. That error comes from https://github.com/shogun-toolbox/shogun/blob/e900a9e602763e1ddc8ac98bb0448ad0a68593e0/src/shogun/base/SGObject.h#L659-L663
Did you write `data->as<DenseSubSamplesFeatures<float64_t>>`?

Reply:
I called `as` function like
```
 if (auto* subfeat_data = data->as<DenseSubSamplesFeatures<float64_t>>(data))
```
I don't know whether this way is right.

Reply:
ah yes, the ->as overload you want to call doesn't have a parameter, so it should be
`data->as<DenseSubSamplesFeatures<float64_t>>()`

Reply:
I change to this way
``` cpp
 if (auto subfeat_data = data->as<DenseSubSamplesFeatures<float64_t>>())
```
But, it can't pass unit test, a exception `C++ exception with description "Object of type DenseFeatures cannot be converted to type shogun::DenseSubSamplesFeatures<double>." thrown in the test body.` is thrown.
I think we should not use `as` function in this place, because if `std::shared_ptr<Features> data` can't converted to `std::shared_ptr<DenseSubSamplesFeatures<double>>`, we should not thrown a exception, we should handle this situation in else clause.

Reply:
ah yes, that's true!